### PR TITLE
[WIP] Add filtering functionality

### PR DESCRIPTION
**Pull request by selene-huang**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
- Added sort by and filtering functionality in the Transactions screen
- Added sort by functionality in the Transaction Details screen
- Implemented bottom modals on the Transactions screen
- Updated Transactions screen to match v2 hifi designs

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy!"
image
image

## How to review
[//]: # "Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?"
Run `npx expo start` and navigate to the Invoices tab.

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

### Currently Not Implemented
- Although the backend is available to sort vouchers on the Transaction Details screen, I couldn't tell how I was supposed to implement the frontend changes to interact with this functionality on the v2 designs @daniellenwong 
- Under the Filters modal, I wasn't sure whether we would prefer to have users input the amount by hand or select the amount using some kind of dropdown, so this is currently not implemented as well

### Known Issues
- The Transactions screen is unable to display transactions when it is first accessed
    - This can be remedied by selecting applying any sort or filter option
    - This cannot be remedied by pulling down to refresh
- Under "Filter by date" in the Filter modal, opening the date picker and pressing "ok" causes the date picker to flash briefly before disappearing
    - Canceling the action does not produce this bug
- The Sort by button does not properly update the positioning of the text when a new sort is applied, as can be seen in the first screenshot.
- When a new sort is applied, the transactions rearrange themselves into some kind of "default" order. You MUST pull to refresh in order to see the properly sorted order.
    - I have not actually figured out what this "default" order is, and why it displays in place of the actual sorted order
    - Related: transactions are supposed to display by date desc as the default (before any sort is applied), but you can only see this AFTER doing something to remedy the first bullet point, and then pulling to refresh

And this part isn't a technical issue, but more of a styling one--for the Filter modal, I couldn't figure out how to style the input fields to
1. display in two columns
2. have the same gap between them in all rows
3. take up the entire row aside from the gap

I guess I could do some kind of screen size math, but that seems less than ideal.

### Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

Updating Transactions and Transaction Details screens to v2 versions from #54 

### Useful Links

https://stackoverflow.com/questions/53048495/does-react-batch-state-update-functions-when-using-hooks
https://github.com/react-native-modal/react-native-modal/issues/268

[//]: # "This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation."
🥦 CC: @sauhardjain

### CODE REVIEW

File: src/screens/transactions/TransactionsScreen.tsx [hunk 1]
```diff
@@ -8,27 +8,59 @@ import { getTransactionsByVendorUuid } from '../../database/queries';
...
+    defaultTransactions,
+    setTransactions,
+  );
+
   const fetchData = async (Uuid: string | null) => {
     try {
       if (Uuid) {
         const transactionsArray = await getTransactionsByVendorUuid(Uuid);
-        setTransactions(transactionsArray);
+        setDefaultTransactions(transactionsArray);
```
Comment (ronniebeggs) [inline src/screens/transactions/TransactionsScreen.tsx:64]: When the page is first accessed, the defaultTransactions list is updated but not rendered. Would it make more sense to automatically render unfiltered/unsorted transactions automatically? Specifically by using setTransactions for the default values instead of setDefaultTransactions.

File: src/utils/transactionUtils.ts [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+            case SortTransactionOption.AMOUNT_ASC:
+              sortedArray = sortTransactionsByAmountAsc(transactions);
+              break;
+            case SortTransactionOption.AMOUNT_DESC:
+              sortedArray = sortTransactionsByAmountDesc(transactions);
+              break;
+            default:
+              sortedArray = sortTransactionsByAmountDesc(transactions);
+          }
+          setTransactions(sortedArray);
```
Comment (ronniebeggs) [inline src/utils/transactionUtils.ts:193]: Here, transactions are sorted and the transactions state variables are updated. Should we create a callback on the TransactionsScreen so the page is re-rendered properly after filtering/sorting? Currently, the only trigger for re-rendering the transactions is the onRefresh callback.

File: src/components/transactions/FilterModal.tsx [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+              setIsVisible(false);
+              filterDispatch({ type: 'ON_SUBMIT' });
+              sortDispatch({ type: 'ON_RELOAD' });
+            }}
+          >
+            <ButtonTextWhite>{`Apply${
+              filterState.inProgressFilterCount > 0
+                ? ` (${filterState.inProgressFilterCount})`
+                : ''
+            }`}</ButtonTextWhite>
```
Comment (ronniebeggs) [inline src/components/transactions/FilterModal.tsx]: The height of the "Apply" button appears to grow as more filters are applied/changes are made.

File: src/components/transactions/SortModal.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        <SortVerticalSpacing />
+        <CenteredContainer>
+          <ButtonMagenta
+            onPress={() => {
+              setIsVisible(false);
+              sortDispatch({ type: 'ON_SUBMIT' });
+            }}
+          >
+            <ButtonTextWhite>Apply</ButtonTextWhite>
+          </ButtonMagenta>
```
Comment (ronniebeggs) [inline src/components/transactions/SortModal.tsx]: A similar issue where the apply button grows as sorting options are selected. Only grows larger with each new button press.

File: src/components/transactions/FilterModal.tsx [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+                  {filterState.inProgressStatusFilter !==
+                  TransactionStatus.UNPAID ? (
+                    <MidGrayText>UNPAID</MidGrayText>
+                  ) : (
+                    'UNPAID'
+                  )}
+                </Body1Text>
+              </CenteredContainer>
+            </View>
+          </FilterField>
```
Comment (ronniebeggs) [inline src/components/transactions/FilterModal.tsx]: Minor style change, but it could be nice for the unpaid/paid filter buttons to match their corresponding color on the transactions screen.

File: src/components/transactions/FilterModal.tsx [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+              if (e.type === 'set' && e.nativeEvent.timestamp) {
+                filterDispatch({
+                  type: 'SET_MAX_DATE',
+                  date: new Date(e.nativeEvent.timestamp),
+                });
+              }
+              setMaxDatePickerIsVisible(false);
+            }}
+          />
+        ) : null}
```
Comment (ronniebeggs) [inline src/components/transactions/FilterModal.tsx]: On iOS, separate date selection buttons appear when the filter by date "From" and "To" options are selected. You can then press these buttons and select dates after they show up. Would it be possible to trigger the data selection modals when our custom button is pressed?

Comment (ronniebeggs) [reply src/components/transactions/FilterModal.tsx]: Also, the clear button doesn't seem to clear date selections.

**Comment by ronniebeggs** [comment]
Overall, it's coming together nicely. ++ for taking on such a large task. You seem to understand most of the issues well. I left some comments but I think they should be resolved fairly easily in time.

**Review by ronniebeggs — CHANGES_REQUESTED** [review]
